pygmt/helpers/utils.py

```diff
@@ -312,7 +316,7 @@ def data_kind(
             # geopandas.GeoDataFrame or shapely.geometry).
             # Reference: https://gist.github.com/sgillies/2217756
             kind = "geojson"
-        case x if x is not None:  # Any not-None is considered as a matrix.
+        case np.ndarray() if data.ndim == 2:  # A 2-D numpy.ndarray object.
             kind = "matrix"
```

Comment:
Two things:

1. Should we extend `kind="matrix"` to 3-D numpy.ndarray objects?
2. I'm hesitant to match based on `np.ndarray()` class only, because while it's not widely advertised, checking for `data is not None` would have caught other array types implementing the `__array_function__` protocol (see [NEP18](https://numpy.org/neps/nep-0018-array-function-protocol.html)), but checking for `isinstance(data, np.ndarray)` would exclude those. That said, I'm wondering if the fallback to `kind="vectors"` might just work, but need to test this out more extensively.

Reply:
> * Should we extend `kind="matrix"` to 3-D numpy.ndarray objects?

It's unclear if we can pass a 3-D numpy array yet. Even if we can, it means more work, since in `virtualfile_from_vectors`/`virtualfile_from_matrix`, we explicitly check if the ndim=1 or 2. So, better to revisit the 3-D numpy array support in a separate PR if necessary.

> * I'm hesitant to match based on `np.ndarray()` class only, because while it's not widely advertised, checking for `data is not None` would have caught other array types implementing the `__array_function__` protocol (see [NEP18](https://numpy.org/neps/nep-0018-array-function-protocol.html)), but checking for `isinstance(data, np.ndarray)` would exclude those. That said, I'm wondering if the fallback to `kind="vectors"` might just work, but need to test this out more extensively.

Yes, I was thinking about checking `__array_interface__` in https://github.com/GenericMappingTools/pygmt/pull/3351#discussion_r1703130372.

Reply:
Done in 4a4f192d71f2754306a80858b6789c93d2fefd08, although other array-like objects are not tested.